python/_tskitmodule.c

```diff
@@ -6632,6 +6632,133 @@ TableCollection_union(TableCollection *self, PyObject *args, PyObject *kwds) {
+
+    ind = 0;
+    for (j = 0; j < num_samples - 1; j++) {
+        for (k = j + 1; k < num_samples; k++) {
+            /* For each pair we return an array of left, right, node values */
+            num_segments = 0;
+            for (u = ibd_finder->ibd_segments_head[ind]; u != NULL; u = u->next) {
+                num_segments++;
+            }
+            left_array = (PyArrayObject *) PyArray_SimpleNew(1, &num_segments, NPY_FLOAT64);
+            right_array = (PyArrayObject *) PyArray_SimpleNew(1, &num_segments, NPY_FLOAT64);
+            node_array = (PyArrayObject *) PyArray_SimpleNew(1, &num_segments, NPY_INT32);
+            if (left_array == NULL || right_array == NULL || node_array == NULL) {
+                goto out;
+            }
+            left = (double *) PyArray_DATA(left_array);
+            right = (double *) PyArray_DATA(right_array);
+            node = (tsk_id_t *) PyArray_DATA(node_array);
+            seg_index = 0;
+            for (u = ibd_finder->ibd_segments_head[ind]; u != NULL; u = u->next) {
+                left[seg_index] = u->left;
+                right[seg_index] = u->right;
+                node[seg_index] = u->node;
+                seg_index++;
+            }
+            key = Py_BuildValue("(ii)", samples[j], samples[k]);
+            value = Py_BuildValue("{s:O,s:O,s:O}",
+                "left", left_array, "right", right_array, "node", node_array);
+            if (key == NULL || value == NULL) {
+                goto out;
+            }
+            if (PyDict_SetItem(pair_dict, key, value) != 0) {
+                goto out;
+            }
+            Py_DECREF(key);
+            Py_DECREF(value);
+            Py_DECREF(left_array);
+            Py_DECREF(right_array);
+            Py_DECREF(node_array);
+            key = NULL;
+            value = NULL;
+            left_array = NULL;
+            right_array = NULL;
+            node_array = NULL;
+            ind++;
+        }
+    }
+
+    ret = pair_dict;
+    pair_dict = NULL;
+out:
+    Py_XDECREF(key);
+    Py_XDECREF(value);
+    Py_XDECREF(left_array);
+    Py_XDECREF(right_array);
+    Py_XDECREF(node_array);
+    Py_XDECREF(pair_dict);
+    return ret;
+}
+
+static PyObject *
+TableCollection_find_ibd(TableCollection *self, PyObject *args, PyObject *kwds)
+{
+    int err;
+    PyObject *ret = NULL;
+    tsk_ibd_finder_t ibd_finder;
+    PyObject *samples;
+    PyArrayObject *samples_array = NULL;
+    double min_length = 0;
+    double max_time = DBL_MAX;
+    npy_intp *shape;
+    static char *kwlist[] = {"samples", "min_length", "max_time", NULL};
+
+    memset(&ibd_finder, 0, sizeof(ibd_finder));
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "O|dd", kwlist, &samples,
+                &min_length, &max_time)) {
+        goto out;
+    }
+    /* NOTE: This will be a 2D array when we read in the pairs explicity */
+    samples_array = (PyArrayObject *) PyArray_FROMANY(samples, NPY_INT32,
```

Comment:
@jeromekelleher I changed this to be a 2D array -- I needed to do this to ensure the low level tests passed the CI.